# Lending Club Loan Performance & Risk Analysis (2007–2018)

Peer-to-peer lending platforms must carefully evaluate borrower risk while maintaining profitability. Poor credit risk assessment can lead to high default rates, reducing investor confidence and financial sustainability.

Lending companies need to understand which borrower characteristics, loan attributes, and economic factors are associated with higher default rates and lower loan performance.

Using historical lending data from Lending Club (2007-2018), this project analyzes loan performance to identify patterns in borrower behavior, loan characteristics, and repayment outcomes. The goal is to generate actionable insights that can help lending companies:

- Identify high-risk borrower profiles

- Understand the relationship between loan characteristics and default rates

- Evaluate the profitability of different loan segments

- Improve lending strategies and risk management decisions

The analysis uses Python for data exploration and cleaning, SQL for business queries, and Power BI to build an interactive dashboard for financial insights.

## Importing dataset

I downloaded the full dataset but for knowledge purposes I'll simulate a SQL database extraction and different SQL queries.

In [2]:
import pandas as pd
import sqlite3

conn = sqlite3.connect("data/lending_analysis.db")

def run_query(query):
    return pd.read_sql(query, conn)

In [3]:
query = """
SELECT *
FROM loans
LIMIT 5;
"""

pd.set_option("display.max_columns", None)
run_query(query)

,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_title,emp_length,home_ownership,annual_inc,verification_status,issue_d,loan_status,pymnt_plan,url,desc,purpose,title,zip_code,addr_state,dti,delinq_2yrs,earliest_cr_line,fico_range_low,fico_range_high,inq_last_6mths,mths_since_last_delinq,mths_since_last_record,open_acc,pub_rec,revol_bal,revol_util,total_acc,initial_list_status,out_prncp,out_prncp_inv,total_pymnt,total_pymnt_inv,total_rec_prncp,total_rec_int,total_rec_late_fee,recoveries,collection_recovery_fee,last_pymnt_d,last_pymnt_amnt,next_pymnt_d,last_credit_pull_d,last_fico_range_high,last_fico_range_low,collections_12_mths_ex_med,mths_since_last_major_derog,policy_code,application_type,annual_inc_joint,dti_joint,verification_status_joint,acc_now_delinq,tot_coll_amt,tot_cur_bal,open_acc_6m,open_act_il,open_il_12m,open_il_24m,mths_since_rcnt_il,total_bal_il,il_util,open_rv_12m,open_rv_24m,max_bal_bc,all_util,total_rev_hi_lim,inq_fi,total_cu_tl,inq_last_12m,acc_open_past_24mths,avg_cur_bal,bc_open_to_buy,bc_util,chargeoff_within_12_mths,delinq_amnt,mo_sin_old_il_acct,mo_sin_old_rev_tl_op,mo_sin_rcnt_rev_tl_op,mo_sin_rcnt_tl,mort_acc,mths_since_recent_bc,mths_since_recent_bc_dlq,mths_since_recent_inq,mths_since_recent_revol_delinq,num_accts_ever_120_pd,num_actv_bc_tl,num_actv_rev_tl,num_bc_sats,num_bc_tl,num_il_tl,num_op_rev_tl,num_rev_accts,num_rev_tl_bal_gt_0,num_sats,num_tl_120dpd_2m,num_tl_30dpd,num_tl_90g_dpd_24m,num_tl_op_past_12m,pct_tl_nvr_dlq,percent_bc_gt_75,pub_rec_bankruptcies,tax_liens,tot_hi_cred_lim,total_bal_ex_mort,total_bc_limit,total_il_high_credit_limit,revol_bal_joint,sec_app_fico_range_low,sec_app_fico_range_high,sec_app_earliest_cr_line,sec_app_inq_last_6mths,sec_app_mort_acc,sec_app_open_acc,sec_app_revol_util,sec_app_open_act_il,sec_app_num_rev_accts,sec_app_chargeoff_within_12_mths,sec_app_collections_12_mths_ex_med,sec_app_mths_since_last_major_derog,hardship_flag,hardship_type,hardship_reason,hardship_status,deferral_term,hardship_amount,hardship_start_date,hardship_end_date,payment_plan_start_date,hardship_length,hardship_dpd,hardship_loan_status,orig_projected_additional_accrued_interest,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
0,68407277,None,3600.0,3600.0,3600.0,36 months,13.99,123.03,C,C4,leadman,10+ years,MORTGAGE,55000.0,Not Verified,Dec-2015,Fully Paid,n,https://lendingclub.com/browse/loanDetail.acti...,None,debt_consolidation,Debt consolidation,190xx,PA,5.91,0.0,Aug-2003,675.0,679.0,1.0,30.0,None,7.0,0.0,2765.0,29.7,13.0,w,0.00,0.00,4421.723917,4421.72,3600.00,821.72,0.0,0.0,0.0,Jan-2019,122.67,NaN,Mar-2019,564.0,560.0,0.0,30.0,1.0,Individual,NaN,NaN,NaN,0.0,722.0,144904.0,2.0,2.0,0.0,1.0,21.0,4981.0,36.0,3.0,3.0,722.0,34.0,9300.0,3.0,1.0,4.0,4.0,20701.0,1506.0,37.2,0.0,0.0,148.0,128.0,3.0,3.0,1.0,4.0,69.0,4.0,69.0,2.0,2.0,4.0,2.0,5.0,3.0,4.0,9.0,4.0,7.0,0.0,0.0,0.0,3.0,76.9,0.0,0.0,0.0,178050.0,7746.0,2400.0,13734.0,None,None,None,None,None,None,None,None,None,None,None,None,None,N,None,None,None,None,None,None,None,None,None,None,None,None,None,None,Cash,N,None,None,None,None,None,None
1,68355089,None,24700.0,24700.0,24700.0,36 months,11.99,820.28,C,C1,Engineer,10+ years,MORTGAGE,65000.0,Not Verified,Dec-2015,Fully Paid,n,https://lendingclub.com/browse/loanDetail.acti...,None,small_business,Business,577xx,SD,16.06,1.0,Dec-1999,715.0,719.0,4.0,6.0,None,22.0,0.0,21470.0,19.2,38.0,w,0.00,0.00,25679.660000,25679.66,24700.00,979.66,0.0,0.0,0.0,Jun-2016,926.35,NaN,Mar-2019,699.0,695.0,0.0,NaN,1.0,Individual,NaN,NaN,NaN,0.0,0.0,204396.0,1.0,1.0,0.0,1.0,19.0,18005.0,73.0,2.0,3.0,6472.0,29.0,111800.0,0.0,0.0,6.0,4.0,9733.0,57830.0,27.1,0.0,0.0,113.0,192.0,2.0,2.0,4.0,2.0,NaN,0.0,6.0,0.0,5.0,5.0,13.0,17.0,6.0,20.0,27.0,5.0,22.0,0.0,0.0,0.0,2.0,97.4,7.7,0.0,0.0,314017.0,39475.

## Data Overview

Prior to any detailed analysis it’s good practice to inspect the raw table.  
In this section we

- report the **shape** (rows × columns).
- review each column’s **dtype**.
- compute basic **summary statistics** for numeric fields.

These checks provide a quick snapshot of data quality and help guide the cleaning and exploratory steps that follow.

### Shape

In [7]:
query = """
SELECT COUNT(*)
FROM loans;
"""

run_query(query)

,COUNT(*)
0,2260701


More than 2 millions of rows is an excellent dataset to analyze and review, however, data may have missing or incorrect values.

### Dtype

In [8]:
query = """
PRAGMA table_info(loans);
"""

run_query(query)

,cid,name,type,notnull,dflt_value,pk
0,0,id,INTEGER,0,None,0
1,1,member_id,REAL,0,None,0
2,2,loan_amnt,REAL,0,None,0
3,3,funded_amnt,REAL,0,None,0
4,4,funded_amnt_inv,REAL,0,None,0
...,...,...,...,...,...,...
146,146,settlement_status,TEXT,0,None,0
147,147,settlement_date,TEXT,0,None,0
148,148,settlement_amount,REAL,0,None,0
149,149,settlement_percentage,REAL,0,None,0


Since the number of columns is 151, reducing it to the most important features is a good approach to focus the analysis in the most reliable columns.

### Summary statistics

## Data Cleaning

Check for missing values, duplicates, and data type issues. Clean the data to ensure accuracy in analysis.

In [ ]:
# Check for missing values
missing_values = df.isnull().sum()
missing_percentage = (missing_values / len(df)) * 100
missing_summary = pd.DataFrame({'Missing Count': missing_values, 'Missing %': missing_percentage})
print("Missing values summary:")
missing_summary[missing_summary['Missing Count'] > 0].sort_values('Missing %', ascending=False)

# Check for duplicates
duplicates = df.duplicated().sum()
print(f"\nNumber of duplicate rows: {duplicates}")

# Data type conversions if needed (e.g., dates)
# df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'], errors='coerce')

## Exploratory Data Analysis (EDA)

Explore the data through visualizations and statistical summaries to uncover patterns, relationships, and insights.

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

# Distribution of loan amounts
plt.figure(figsize=(10, 6))
sns.histplot(df['loan_amnt'], bins=50, kde=True)
plt.title('Distribution of Loan Amounts')
plt.xlabel('Loan Amount ($)')
plt.ylabel('Frequency')
plt.show()

# Loan status by grade
plt.figure(figsize=(10, 6))
sns.countplot(data=df, x='grade', hue='loan_status', order=sorted(df['grade'].unique()))
plt.title('Loan Status by Grade')
plt.show()

# Correlation heatmap for numerical variables
numerical_cols = ['loan_amnt', 'int_rate', 'annual_inc', 'dti', 'fico_range_low', 'fico_range_high']
corr_matrix = df[numerical_cols].corr()
plt.figure(figsize=(10, 8))
sns.heatmap(corr_matrix, annot=True, cmap='coolwarm', fmt='.2f')
plt.title('Correlation Matrix')
plt.show()

## Key Insights and Recommendations

Summarize the main findings from the analysis and provide actionable recommendations for Lending Club.

### Recommendations for a Perfect Data Analyst Notebook

1. **Structure with Clear Sections**: Use markdown headers to organize the notebook into logical sections (e.g., Import, Overview, Cleaning, EDA, Insights).

2. **Document Everything**: Add comments in code cells and explanatory text in markdown cells to explain why you're doing each step and what the results mean.

3. **Handle Data Quality**: Always check for missing values, outliers, and data types. Decide on imputation or removal strategies based on business context.

4. **Use Visualizations**: Include charts and graphs to make insights visual and easier to understand. Label axes and add titles.

5. **Focus on Business Value**: Tie analysis back to the original questions (e.g., which customers are likely to default, profitability by segment).

6. **Ensure Reproducibility**: Include environment setup (library versions), and make sure the notebook runs from top to bottom.

7. **Summarize Key Findings**: End with a concise summary of insights and recommendations, avoiding technical jargon for non-technical audiences.

8. **Version Control**: Use Git to track changes, and consider exporting to PDF or HTML for sharing.

9. **Performance Considerations**: For large datasets, use sampling or efficient queries to avoid long run times.

10. **Ethical Considerations**: Be mindful of data privacy and bias in analysis, especially with financial data.

## Loan status distribution

In [4]:
query = """
SELECT 
    loan_status,
    COUNT(*) AS total_loans,
    ROUND(COUNT(*) * 100.0 / (
        SELECT COUNT(*) 
        FROM loans 
        WHERE loan_status IS NOT NULL
    ), 2) AS percentage
FROM loans
WHERE loan_status IS NOT NULL
GROUP BY loan_status
ORDER BY total_loans DESC;
"""

total_loans_by_status = run_query(query)
total_loans_by_status

,loan_status,total_loans,percentage
0,Fully Paid,1076751,47.63
1,Current,878317,38.85
2,Charged Off,268559,11.88
3,Late (31-120 days),21467,0.95
4,In Grace Period,8436,0.37
5,Late (16-30 days),4349,0.19
6,Does not meet the credit policy. Status:Fully ...,1988,0.09
7,Does not meet the credit policy. Status:Charge...,761,0.03
8,Default,40,0.00


1. **High Repayment Success**: Nearly half (47.63%) of all loans are fully paid, indicating that Lending Club's borrower selection and risk assessment are effective for a significant portion of applicants.

2. **Significant Ongoing Exposure**: Current loans account for 38.85% of the total, representing active accounts that could still transition to fully paid or charged off, requiring continuous monitoring.

3. **Moderate Loss Rate**: Charged-off loans make up 11.88% of the total, a notable but manageable portion of financial losses that highlights areas for improving default prevention.

## Average loan amount by loan status

In [6]:
query = """
SELECT 
    loan_status,
    AVG(loan_amnt) AS avg_loan_amount
FROM loans
GROUP BY loan_status
ORDER BY avg_loan_amount DESC;
"""

average_loan_amount_by_loan_status = run_query(query)
average_loan_amount_by_loan_status

,loan_status,avg_loan_amount
0,In Grace Period,17672.558084
1,Late (16-30 days),17391.118648
2,Late (31-120 days),16946.614571
3,Current,15942.815920
4,Charged Off,15565.055444
5,Default,14350.625000
6,Fully Paid,14134.369808
7,Does not meet the credit policy. Status:Charge...,9527.233903
8,Does not meet the credit policy. Status:Fully ...,8853.231891
9,NaN,NaN


In [6]:
query = """
SELECT
    -- Target variable
    loan_status,

    -- Credit profile
    fico_range_low, 
    fico_range_high, 
    grade, 
    sub_grade,
    delinq_2yrs, 
    inq_last_6mths,

    -- Debt & Income metrics
    dti, 
    annual_inc, 
    revol_util,

    -- Loan specifics
    loan_amnt, 
    int_rate, 
    term, 
    purpose,

    -- Credit history
    earliest_cr_line, 
    total_acc, 
    open_acc,
    mths_since_last_delinq, 
    mths_since_last_major_derog

FROM loans;
"""

df = run_query(query)